engine/runtime/src/main/java/org/enso/interpreter/runtime/scope/TopLevelScope.java

```diff
@@ -233,6 +241,83 @@ private static Object compile(Object[] arguments, EnsoContext context)
       }
     }
 
+    private static boolean compileMultipleProjects(Object[] args, EnsoContext ctx) {
+      assert args.length == 3;
+      var shouldCompileDependencies = Boolean.TRUE.equals(args[0]);
+      Option<String> generateDocs =
+          switch (args[1]) {
+            case Boolean b when !b -> Option.empty();
+            case String s -> Option.apply(s);
+            default -> Option.empty();
+          };
+      var projectPaths = asInteropStringList(args[2]);
+      return compileMultipleProjects(projectPaths, ctx, generateDocs, shouldCompileDependencies);
+    }
+
+    /**
+     * Compiles multiple projects one by one. Every project is registered as the main project via
+     * {@link PackageRepository#registerMainProjectPackage} and then it is compiled.
+     *
+     * <p>Note that it is currently not possible to run the compilation in parallel.
+     *
+     * @param projPaths Paths to the project directories.
+     * @return False if one of the compilations failed, true if all succeeded.
+     */
+    private static boolean compileMultipleProjects(
+        List<String> projPaths,
+        EnsoContext ctx,
+        Option<String> generateDocs,
+        boolean shouldCompileDependencies) {
+      var fs = TruffleFileSystem.INSTANCE;
+      var pkgManager = new PackageManager<>(fs);
+      var shouldWriteCache = !ctx.isIrCachingDisabled();
+      for (var projPath : projPaths) {
+        var dir = ctx.getPublicTruffleFile(projPath);
+        var loadedPkg =
+            pkgManager
+                .loadPackage(dir)
+                .fold(
+                    err -> {
+                      throw new ProjectLoadingFailure(dir.getName(), err);
+                    },
+                    res -> res);
+        ctx.getPackageRepository().registerMainProjectPackage(loadedPkg.libraryName(), loadedPkg);
+        var fut =
+            ctx.getCompiler().compile(shouldCompileDependencies, shouldWriteCache, generateDocs);
+        boolean compilationSucceeded;
+        try {
+          compilationSucceeded = fut.get();
+        } catch (InterruptedException e) {
+          throw new RuntimeException(e);
+        } catch (ExecutionException e) {
+          var re = new RuntimeException(e);
+          re.setStackTrace(e.getStackTrace());
+          throw re;
+        }
+        if (!compilationSucceeded) {
+          return false;
+        }
+      }
+      return true;
+    }
+
+    private static List<String> asInteropStringList(Object obj) {
```

Comment:
The name seems upside down:
```suggestion
    private static List<String> fromInteropStringList(Object obj) {
```
isn't the input the `hasArrayElements(isString(...))` interop object and the output is `java.util.List`?

Reply:
Yes, it is upside down. Renamed in 29792a216677db9369dc92a5ae7924e42d6a90f3